libmcount/mcount.c

```diff
@@ -1716,6 +1724,145 @@ static void mcount_script_init(enum uftrace_pattern_type patt_type)
 	strv_free(&info.cmds);
 }
 
+/* unlink socket file if it exists */
+static void unlink_socket_maybe(struct sockaddr_un *addr)
+{
+	if (unlink(addr->sun_path) == -1) {
+		if (errno != ENOENT) {
+			pr_dbg("cannot unlink %s\n", addr->sun_path);
+		}
+	}
+}
+
+static int agent_init(struct sockaddr_un *addr)
+{
+	int sfd;
+
+	if (mkdir(MCOUNT_AGENT_SOCKET_DIR, 0775) == -1) {
+		if (errno != EEXIST) {
+			pr_dbg("error creating run directory %s\n", MCOUNT_AGENT_SOCKET_DIR);
+		}
+	}
+
+	sfd = socket_create(addr, getpid());
+	if (sfd == -1)
+		return sfd;
+
+	unlink_socket_maybe(addr);
+
+	if (socket_listen(sfd, addr) == -1)
+		return -1;
+
+	return sfd;
+}
+
+static void agent_fini(struct sockaddr_un *addr, int sfd)
+{
+	if (sfd != -1)
+		close(sfd);
+
+	unlink_socket_maybe(addr);
+}
+
+/* Agent routine, applying instructions from the CLI. */
+void *agent_apply_commands(void *arg)
+{
+	int sfd, cfd; /* socket fd, connection fd */
+	bool close_connection;
+	enum uftrace_dopt dopt;
+	struct sockaddr_un addr;
+
+	sfd = agent_init(&addr);
+	if (sfd == -1) {
+		pr_warn("agent cannot start\n");
+		return NULL;
+	}
+
+	agent_run = true;
+	pr_dbg("agent started on socket %s\n", addr.sun_path);
+
+	while (agent_run) {
+		cfd = socket_accept(sfd);
+		if (cfd == -1) {
+			pr_warn("error accepting socket connection\n");
+			continue;
+		}
+		pr_dbg2("agent connection open\n");
+
+		close_connection = false;
+		while (!close_connection) {
+			if (read(cfd, &dopt, sizeof(enum uftrace_dopt)) == -1) {
+				pr_warn("error reading option\n", errno);
+				break;
+			}
+
+			switch (dopt) {
+			case UFTRACE_DOPT_CLOSE:
+				close_connection = true;
+				break;
+
+			default:
+				close_connection = true;
+				pr_warn("option not recognized: %d\n", dopt);
+			}
+		}
+		if (close(cfd) == -1)
+			pr_dbg2("cannot close socket connection\n");
+		else
+			pr_dbg2("agent connection closed\n");
+	}
+
+	agent_fini(&addr, sfd);
+
+	pr_dbg("agent exited\n");
+	return 0;
+}
+
+static void agent_spawn()
+{
+	errno = pthread_create(&agent, NULL, &agent_apply_commands, NULL);
+	if (errno != 0)
+		pr_warn("cannot start agent: %s\n", strerror((errno)));
+}
+
+/* Check if the agent is up. If so, set its run flag to false, open and close
+ * connection . */
+static void agent_kill()
+{
+	int sfd;
+	struct sockaddr_un addr;
+
+	if (agent_run)
+		agent_run = false;
+	else
+		return;
+
+	sfd = socket_create(&addr, getpid());
+	if (sfd == -1)
+		goto error;
+
+	if (socket_connect(sfd, &addr) == -1) {
+		if (errno != ENOENT) /* The agent may have ended and deleted the socket */
+			goto error;
+	}
+
+	if (socket_send_option(sfd, UFTRACE_DOPT_CLOSE, NULL, 0) == -1)
+		pr_dbg("cannot stop agent loop\n");
+
+	close(sfd);
+
+	switch (pthread_join(agent, NULL)) {
+	case 0:
+	case EINVAL:
+		break;
+	default:
+		pr_dbg("agent left in unknown state\n");
+	}
```

Comment:
Why do you handle EINVAL specially?  I think we can just add `pr_dbg()` for the any error return.

Reply:
It was in case the agent thread terminated by itself, but this shouldn't happen.